File: src/spkmeans.c
#include "spkmeans.h"
#include "myIO.h"
#include "calculate_w.h"
#include "lnorm.h"
#include "eigengap.h"
#include <stdio.h>
#include <stdlib.h>

#define ERROR_MSG "An Error Has Occurred"
#define INVALID_INPUT_MSG "Invalid Input!"
#define terminate(msg)      do { printf("%s\n", msg); exit(1); } while(0)

int run(GOAL goal, const char *file_name) {
    MATRIX X = read_vectors_from_file(file_name);
    if (is_null(X)) return 1;
    if (goal == JACOBI) {
        /* TODO run Jacobi algo */
    }
    else /* goal in {WAM, DDG, LNORM} */ {
        MATRIX W, L; DIAGONAL_MATRIX D;
        W = /* TODO func tha clacs W */ NULL_MATRIX;
        if (is_null(W)) return 1;
        if (goal == WAM) {
            print_matrix(W);
            return 0;
        }
        D = diagonal_degree_matrix(W);
        if (is_null(D)) return 1;
        if (goal == DDG) {
            print_diagonal_matrix(D);
            return 0;
        }
        L = normalized_graph_laplacian(W, D);
        if (is_null(L)) return 1;
        if (goal == LNORM) {
            print_matrix(L);
            return 0;
        }
    }
    return 2;
}

MATRIX spectralization(const char *file_name, unsigned k) {
    MATRIX X, W, L, T; DIAGONAL_MATRIX D; MAT_AND_VEC U; 

    X = read_vectors_from_file(file_name);
    if (is_null(X)) return NULL_MATRIX;
    /* TODO calc W */
    if (is_null(W)) return NULL_MATRIX;
    D = diagonal_degree_matrix(W);
    if (is_null(D)) return NULL_MATRIX;
    L = normalized_graph_laplacian(W, D);\
    if (is_null(L)) return NULL_MATRIX;
    /* TODO jacobi */
    if (k > 0) sort_inplace(&U);
    else k = k_and_sort_inplace(&U);
    T = reduced_vectors(U.matrix, k);
    if (is_null(T)) return NULL_MATRIX;
    return T;
}

int main (int argc, char **argv) {
    if (argc != 3) terminate(INVALID_INPUT_MSG);
    if (!has_input_file_name_format(argv[2])) terminate(INVALID_INPUT_MSG);
    /* TODO find goal */
    return 0;
}

File: src/spkmeans.py
import mykmeanssp

import sys
import numpy as np


INVALID_INPUT_MSG = 'Invalid Input!'
ERROR_MSG = 'An Error Has Occurred'

goals = {
    'spk' : -1,
    'wam' : 0,
    'ddg' : 1,
    'lnorm' : 2,
    'jacobi' : 3,
}


def get_input():
    """returns k, goal, input_file_name"""
    args = sys.argv[1:]
    assert len(args) == 3
    assert args[2].endswith(('.txt', '.csv'))
    k = int(args[0])
    assert k >= 0
    return k, goals[args[1]], args[2]

def kmeans_pp(vectors: np.ndarray):
    """
    runs k-means++ algorithms and returns its result (k = dim.)
    """
    num_vecs, dim = vectors.shape
    np.random.seed(0)
    centroids = [ np.random.choice(num_vecs) ]
    norm_squared = lambda v: np.inner(v, v)

    for _ in range(1, dim):
        P = np.fromiter(
            (
                min(
                    norm_squared(vectors[l] - vectors[j]) for j in centroids
                ) 
                for l in range(num_vecs)
            ), 
            float
        ) 
        P = P / np.sum(P)
        centroids.append(np.random.choice(num_vecs, p=P))
    print(','.join(str(i) for i in centroids))
    return vectors[centroids]


def main():
    try:
        k, goal, input_file_name = get_input()
    except:
        sys.exit(INVALID_INPUT_MSG)
    try:
        if goal == goals['spk']:
            vectors = mykmeanssp.transform(input_file_name, k)
            initial_centroids = kmeans_pp(vectors)
            assert mykmeanssp.kmeans(
                vectors, initial_centroids, *vectors.shape
                ) == 0
        else:
            assert mykmeanssp.run(goal, input_file_name) == 0
    except:
        sys.exit(ERROR_MSG)

if __name__ == '__main__':
    main()